### PR TITLE
Use `etcd-main` endpoint for full snapshots

### PR DESCRIPTION
**Pull request by timuthy**
**How to categorize this PR?**

/area control-plane
/kind technical-debt

**What this PR does / why we need it**:
Gardener used a `PodExecutor` to initiate full etcd snapshots. For performance and simplification reasons, it should rather call the HTTP endpoint of the `etcd-main` service directly. In addition, we drop dependencies on binaries that must be available in the `etcd-backup-restore` container, e.g. it was expected that `curl` was available in the container image.

**Special notes for your reviewer**:
/cc @ishan16696

**Release note**:

```other operator
Gardenlet switched from a "PodExec" based approach to calling the appropriate HTTP endpoint for initiating full snapshots. This usually takes place when the ETCD encryption key was rotated or during control-plane migration.
```

### CODE REVIEW

File: pkg/operation/botanist/component/etcd/etcd.go [hunk 1]
```diff
@@ -679,28 +684,23 @@ func (e *etcd) emptyHVPA() *hvpav1alpha1.Hvpa {
...
 }
 
-func (e *etcd) Snapshot(ctx context.Context, podExecutor kubernetes.PodExecutor) error {
+func (e *etcd) Snapshot(ctx context.Context, httpClient rest.HTTPClient) error {
 	if e.values.BackupConfig == nil {
 		return fmt.Errorf("no backup is configured for this etcd, cannot make a snapshot")
 	}
 
-	etcdMainSelector := e.podLabelSelector()
+	url := fmt.Sprintf("https://etcd-%s-client.%s:%d/snapshot/full?final=true", e.values.Role, e.namespace, etcdconstants.PortBackupRestore)
```
Comment (rfranzke) [inline pkg/operation/botanist/component/etcd/etcd.go]: Please use https://github.com/gardener/gardener/blob/77dea8baa3044a49bd075402e86075def246112f/pkg/operation/botanist/component/etcd/constants/constants.go#L22 instead

File: pkg/operation/botanist/component/etcd/etcd.go [hunk 2]
```diff
@@ -679,28 +684,23 @@ func (e *etcd) emptyHVPA() *hvpav1alpha1.Hvpa {
...
+	request, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
+	if err != nil {
 		return err
 	}
-	if len(podsList.Items) == 0 {
-		return fmt.Errorf("didn't find any pods for selector: %v", etcdMainSelector)
+
+	resp, err := httpClient.Do(request)
+	if err == nil && resp != nil && resp.StatusCode != http.StatusOK {
+		return fmt.Errorf("error occurred while intiating etcd snapshot: %s", resp.Status)
```
Comment (rfranzke) [inline pkg/operation/botanist/component/etcd/etcd.go]:
```suggestion
		return fmt.Errorf("error occurred while initiating ETCD snapshot: %s", resp.Status)
```

**Comment by ishan16696** [comment]
Thanks for PR @timuthy,
I have only 1 doubt: I see others tests like [this](https://github.com/gardener/gardener/blob/8d58fefd1366af651ead68cef104eecaee18e417/test/testmachinery/shoots/applications/networking.go#L90) also uses `PodExecutor` to execute some command. Is this fine ? I know it is out of scope of this PR I'm just asking as if there is a plan to move to distroless base image then tests like [this](https://github.com/gardener/gardener/blob/8d58fefd1366af651ead68cef104eecaee18e417/test/testmachinery/shoots/applications/networking.go#L90)  will also need to revisit.

**Review by ishan16696 — APPROVED** [review]
LGTM!!

**Comment by timuthy** [comment]
@ishan16696 thanks for the hint.

The `PodExecutor` is used in different Test Machinery based tests:
- Loki: https://github.com/gardener/gardener/blob/3b33e4e25302bdfe254b9187b2cfe4c7885e43b9/test/framework/shoot_utils.go#L51
- Nginx: https://github.com/gardener/gardener/blob/3b33e4e25302bdfe254b9187b2cfe4c7885e43b9/test/testmachinery/shoots/applications/networking.go#L79
- VPN Tunnel: https://github.com/gardener/gardener/blob/3b33e4e25302bdfe254b9187b2cfe4c7885e43b9/test/testmachinery/shoots/vpntunnel/vpntunnel.go#L91

Except for Loki, the tests manage their own deployments for the tests, so I don't see a reason to switch to a distroless image there, yet. The tests with Loki can become an issue if there is a general switch to a distroless image.

@nickytd do you know of any plans to switch to a distroless based Loki image?

**Comment by nickytd** [comment]
There are no such plans. Lokis will be replaced with Vali container images https://github.com/credativ/vali/releases/tag/v2.2.4  
And those are alpine based.

**Pull request merged**